### PR TITLE
feat: Add (group) ownership functionality

### LINKED ISSUE #42: Allow specifying owner and group
Currently the remote file ends up being owned by the SSH user. It would be useful, especially in combination with `sudo = true` if the file could be `chown`ed to be owned by another user/group.

### PR DESCRIPTION
**Pull request by bjw-s**
This PR aims to add the ability to specify the owner UID / group GID for a file.

Should fix #42

### CODE REVIEW

File: internal/provider/resource_remote_file.go [hunk 1]
```diff
@@ -74,6 +90,18 @@ func resourceRemoteFileCreate(ctx context.Context, d *schema.ResourceData, meta
...
+				return diag.Errorf("error while changing group of remote file with sudo: %s", err.Error())
+			}
+		}
+		if owner != "" {
+			err = client.ChownFileSudo(path, owner)
+			if err != nil {
+				return diag.Errorf("error while changing owner of remote file with sudo: %s", err.Error())
+			}
+		}
 	} else {
```
Comment (tenstad) [inline internal/provider/resource_remote_file.go:105]: Would like to have ownership functionality for non-`sudo` connections as well. Does not seem like the scp client supports ownership, so I think we should do it with `chown` and `chgrp`. Could you modify `ChgrpFileSudo` and `ChownFileSudo` to take a `sudo: bool` parameter, and only include `sudo` in the command when `sudo` is `true`? We can then reuse the functions for non-`sudo` connections.

Comment (tenstad) [inline internal/provider/resource_remote_file.go:105]: Would like to have ownership functionality for non-`sudo` connections as well. Does not seem like the scp client supports ownership, so I think we should do it with `chown` and `chgrp`. Could you modify `ChgrpFileSudo` and `ChownFileSudo` to take a `sudo: bool` parameter, and only include `sudo` in the command when `sudo` is `true`? We can then reuse the functions for non-`sudo` connections.

**Review by tenstad — COMMENTED** [review]
Thank you for the PR! I would like a test that checks that the ownership is correct, but I can fix it in a future PR if you wish.

**Comment by bjw-s** [comment]
Thanks for the feedback! I'll try to take a crack at the suggestions over the weekend!

File: internal/provider/resource_remote_file.go (hunk 1 above)
Comment (bjw-s) [reply internal/provider/resource_remote_file.go:103]: I've added this in my latest commit

**Comment by bjw-s** [comment]
> Thank you for the PR! I would like a test that checks that the ownership is correct, but I can fix it in a future PR if you wish.

I've added these tests, although I am not 100% certain that this is how you would want them to be implemented. Please let me know if you want me to adjust anything.

**Comment by tenstad** [comment]
Looks good! As with permissions (https://github.com/tenstad/terraform-provider-remote/issues/51), we should read the UID and GID of the file in [resourceRemoteFileRead](https://github.com/tenstad/terraform-provider-remote/blob/d7d7f79db4dc7f3c1cd8a94eeb116732c6661550/internal/provider/resource_remote_file.go#L92), and update the resource. If you wish to attempt it in this PR, that would be great. I will also accept creating a separate issue for it (as with permissions) and having ownership be write-only for now.

**Comment by tenstad** [comment]
> I did have to duplicate quite a bit [..] I am open to making this a bit more DRY, but maybe that can be done in a separate PR?

Yeah, we can clean up a bit in a separate PR :)

Thank you for the additional changes! You have now also fixed https://github.com/tenstad/terraform-provider-remote/issues/51 :D I think we are ready to merge after my single new comment is resolved, good work :+1:

File: internal/provider/resource_remote_file.go [hunk 2]
```diff
@@ -115,6 +156,27 @@ func resourceRemoteFileRead(ctx context.Context, d *schema.ResourceData, meta in
 				return diag.Errorf("error while reading remote file with sudo: %s", err.Error())
 			}
 			d.Set("content", content)
+			if permissions != "" {
+				permissions, err := client.ReadFilePermissions(path, true)
+				if err != nil {
+					return diag.Errorf("error while reading remote file permissions with sudo: %s", err.Error())
+				}
+				d.Set("permissions", permissions)
+			}
```
[the remark this answers is not in the document]
Comment (bjw-s) [reply internal/provider/resource_remote_file.go]: Done in the latest commit :)

**Comment by bjw-s** [comment]
Cool... Look like everything should be good to go now then 🙂

**Review by tenstad — APPROVED** [review]

**Comment by bjw-s** [comment]
Thanks for the PR approval! Just making pure it's intentional that you haven't also merged it? 😇

**Comment by tenstad** [comment]
> Thanks for the PR approval! Just making pure it's intentional that you haven't also merged it? innocent

Yeah, wanted to do a final round of manual testing :smile: Merged now, and will soon create a new release.

**Pull request merged**